File: src/OS/runtime/lifecycle.py
from typing import ClassVar

from OS.domain.agent import AgentInstance, AgentStatus


class LifecycleError(Exception):
    """
    非法生命周期转换。
    """


class LifecycleManager:
    """
    Agent 生命周期管理。
    """

    _transitions: ClassVar[dict[AgentStatus, set[AgentStatus]]] = {
        # 允许的状态变化
        AgentStatus.CREATED: {AgentStatus.ACTIVE},
        AgentStatus.ACTIVE: {AgentStatus.PAUSED, AgentStatus.TERMINATING},
        AgentStatus.PAUSED: {AgentStatus.ACTIVE, AgentStatus.TERMINATING},
        AgentStatus.TERMINATING: {AgentStatus.TERMINATED},
        AgentStatus.TERMINATED: set(),
    }

    def transition(
        self,
        agent: AgentInstance,
        target: AgentStatus,
    ) -> None:
        """
        尝试迁移 Agent 状态。
        """

        allowed = self._transitions.get(agent.status, set())

        if target not in allowed:
            raise LifecycleError(f"不能从{agent.status.value}切换到{target.value}")

        agent.status = target

File: runtime/core/lifecycle.go
package core

import (
	"agent-runtime/domain"
	"fmt"
)

//agent生命周期转换
type LifecycleManager struct{}

var allowedTransitions = map[domain.AgentStatus]map[domain.AgentStatus]bool{
	domain.AgentStatusCreated:     {domain.AgentStatusActive: true},
	domain.AgentStatusActive:      {domain.AgentStatusPaused: true, domain.AgentStatusTerminating: true},
	domain.AgentStatusPaused:      {domain.AgentStatusActive: true, domain.AgentStatusTerminating: true},
	domain.AgentStatusTerminating: {domain.AgentStatusTerminated: true},
	domain.AgentStatusTerminated:  {},
}

//切换状态，非法转换保持原状
func (LifecycleManager) Transition(agent *domain.AgentInstance, target domain.AgentStatus) error {
	if agent == nil {
		return fmt.Errorf("切换 agent 生命周期: agent 不能为空")
	}
	if !allowedTransitions[agent.Status][target] {
		return fmt.Errorf("不能从 %s 切换到 %s", agent.Status, target)
	}
	agent.Status = target
	return nil
}
